src/jit/hwintrinsiccodegenxarch.cpp

```diff
@@ -297,6 +298,262 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
+                    assert(memIndir->Offset() == 0);
+
+                    break;
+                }
+
+                case GT_CLS_VAR_ADDR:
+                {
+                    emit->emitIns_R_C(ins, attr, targetReg, memBase->gtClsVar.gtClsVarHnd, 0);
+                    return;
+                }
+
+                default:
+                {
+                    emit->emitIns_R_A(ins, attr, targetReg, memIndir);
+                    return;
+                }
+            }
+        }
+        else
+        {
+            switch (op1->OperGet())
+            {
+                case GT_LCL_FLD:
+                {
+                    GenTreeLclFld* lclField = op1->AsLclFld();
+
+                    varNum = lclField->GetLclNum();
+                    offset = lclField->gtLclFld.gtLclOffs;
+                    break;
+                }
+
+                case GT_LCL_VAR:
+                {
+                    assert(op1->IsRegOptional() || !compiler->lvaTable[op1->gtLclVar.gtLclNum].lvIsRegCandidate());
+                    varNum = op1->AsLclVar()->GetLclNum();
+                    offset = 0;
+                    break;
+                }
+
+                default:
+                {
+                    unreached();
+                    break;
+                }
+            }
+        }
+
+        // Ensure we got a good varNum and offset.
+        // We also need to check for `tmpDsc != nullptr` since spill temp numbers
+        // are negative and start with -1, which also happens to be BAD_VAR_NUM.
+        assert((varNum != BAD_VAR_NUM) || (tmpDsc != nullptr));
+        assert(offset != (unsigned)-1);
+
+        emit->emitIns_R_S(ins, attr, targetReg, varNum, offset);
+    }
+    else
+    {
+        regNumber op1Reg = op1->gtRegNum;
+        emit->emitIns_R_R(ins, attr, targetReg, op1Reg);
+    }
+}
+
+//------------------------------------------------------------------------
+// genHWIntrinsic_R_RM_I: Generates the code for a hardware intrinsic node that takes a register operand, a
+//                        register/memory operand, an immediate operand, and that returns a value in register
+//
+// Arguments:
+//    node - The hardware intrinsic node
+//    ins  - The instruction being generated
+//
+void CodeGen::genHWIntrinsic_R_RM_I(GenTreeHWIntrinsic* node, instruction ins)
+{
+    var_types targetType = node->TypeGet();
+    regNumber targetReg  = node->gtRegNum;
+    GenTree*  op1        = node->gtGetOp1();
+    emitAttr  simdSize   = EA_ATTR(node->gtSIMDSize);
+    emitter*  emit       = getEmitter();
+
+    int ival = HWIntrinsicInfo::lookupIval(node->gtHWIntrinsicId);
+    assert((ival >= 0) && (ival <= 127));
+
+    // TODO-XArch-CQ: Commutative operations can have op1 be contained
+    // TODO-XArch-CQ: Non-VEX encoded instructions can have both ops contained
+
+    assert(targetReg != REG_NA);
+    assert(node->gtGetOp2() == nullptr);
+    assert(!node->OperIsCommutative());
```

Comment:
Why is this asserted? It would seem to be possible that a 3-operand instruction could potentially be commutative in its first two operands.

Reply:
Yes, the FMA instructions actually fall into this category.

However, they also require special handling as there isn't any clear indication of which operands are commutative.

I'll add a comment indicating this exists because of the above (which should also leave it open to removal of needed later).

Reply:
See https://github.com/dotnet/coreclr/pull/18262#discussion_r192866575 for the incorrect function header. Added a comment anyways indicating one operand intrinsics can't be commutative.